core/chaincode/lifecycle/lifecycle.go

```diff
@@ -675,6 +675,9 @@ func (ef *ExternalFunctions) InstallChaincode(chaincodeInstallPackage []byte) (*
 	}
 
 	buildStatus, ok := ef.BuildRegistry.BuildStatus(packageID)
+	if buildStatus != nil && (buildStatus.Err() != nil) {
+		buildStatus, ok = ef.BuildRegistry.ResetBuildStatus(packageID)
```

Comment:
As pointed out in the comment in the build status tracker, we need to implement internal locking here as well.  We need to make sure that we don't end up getting two resets done after a build has failed.

I'd suggest that we take a lock, based on the package ID, prior to attempting to write the chaincode package to the filesystem.  And released either after everything is successful or the chaincode is removed.  This would protect us against concurrent installs of the same chaincode.